### ISSUE TITLE
issue with handling of empty numeric input boxes

### ISSUE THREAD

**Issue by DenoBeno** [NetLogo/NetLogo#682]
Note: initially I thought the issue way with the "carefully", but it turned out that I wasn't careful enough when reporting the issue. In fact the problem was (is) with error checking on the entry field. 
--- Initial report --
I tried to use the entry fields to set up the model. Since the entry field can be empty, I tried to use "carefully" to catch such cases and set default value if nothing is provided. Something like:

to setup-turtles
  Carefully [  Create-turtles N ] [ 
       Set N 100 
      Create-turtles N
    ]
end

I tested this in a simple model which has an entry field called N. On the first try (N undefined), I get this runtime error: 

ASK expected input to be an agent or agentset but got NOBODY instead.
error while observer running ASK
  called by procedure SETUP-TURTLES
  called by procedure SETUP
  called by Button 'setup'

so basically the program comes to runtime-exception checking part and then fails with runtime exception. Nice. :-)

And it gets better: if I run the model with some number and then delete the number, the error changes into "Invalid input for a number: expected a number" exception instead. Which is probably the runtime exception check on the input form. 

hope this helps

**Comment by SethTisue** [comment]
I'm unable to reproduce the problem on my own computer. I don't think you've provided all of the relevant code? The error says "ASK expected input to be an agent or agentset", but `ask` doesn't appear anywhere in the code you've provided.

**Comment by digitaldust** [comment]
I tried too, but the only thing I can reproduce is that if you delete the
number from the input and then try to call setup-turtles, then you have the
expected "Invalid input for a number: expected a number" exception when you
leave the input widget.

2015-01-07 15:09 GMT+01:00 Seth Tisue <notifications@github.com>:

> I'm unable to reproduce the problem on my own computer. I don't think
> you've provided all of the relevant code? The error says "ASK expected
> input to be an agent or agentset", but ask doesn't appear anywhere in the
> code you've provided.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NetLogo/NetLogo/issues/682#issuecomment-69025902>.
>



-- 
-----------------------------------------------------------------

Simone Gabbriellini, PhD

Post-doctoral Researcher
ANR founded research project "DIFFCERAM"
GEMASS, CNRS & Paris-Sorbonne.

mobile: +39 340 39 75 626
email: simone.gabbriellini@cnrs.fr

**Comment by DenoBeno** [comment]
Sorry, mea culpa. Here is the simplest code I could think of:

to setup
  clear-all
  setup-world
  setup-turtles
  reset-ticks
end

to setup-world
  ask patches [
      set pcolor grey
  ]
end

to setup-turtles
  Carefully [  Create-turtles N ] [ 
    Set N 100 
    Create-turtles N
  ]
  ; first-time error happens here.
  ask one-of turtles [
    pen-down
  ]
end

to go
  ask turtles [ forward 1 ]
end

Other than that, it's just two buttons (setup, go) and one input field (N).

The "pendown" for one turtle was added just for fun, but that's the part of the code which really causes the error. I presume that N variable is set to zero at the start and therefore causes no exceptions. This is reproducible - just delete the "N" input field (type: number) and try again. Later on, the exception is caught by the input field which is kind-of OK.

Thus, the real issue is thus probably not in the "carefully" code, but in the input field which for some reason does not recognise the initial empty entry as the issue.

**Comment by SethTisue** [comment]
An empty numeric input box is considered to have value 0. What behavior are you suggesting instead?

**Comment by DenoBeno** [comment]
Dear Seth,

I understand that NetLogo automatically sets the variable value to zero if not otherwise defined, rather than using NaN. This is a design decision and I don't suggest changing it..

The only thing that bothers me at the end of the day is that the input field for some reason reacts differently when it's initialized (no value, never had any value before) and at a later time, when an existing value has been deleted.

In the first case, the value of zero is assumed, in the second the widget rises an exception. This is confusing.

My suggestion is: 

1) Let me define a function (in simplest case just set a default value) which is called in case an empty field is encountered. 
2) Second best would be to rise the same "value not defined" exception in the both cases. 
3) If none of the above is possible. the system should at least assume zero value in both cases and let me handle the situation. I do not like this solution, because zero may be a valid input and therefore I need to differ between "zero value" and "empty field".

hope this helps,

 Denis

**Comment by SethTisue** [comment]
I don't think there was ever an conscious intention to allow a numeric input box to ever be empty at all. I think it's simply an oversight that when you create the input box, it begins as empty rather than explicitly containing "0". So just fixing the implementation to match that intention is another possible path here. (But perhaps the maintainers will also consider making one of your somewhat more elaborate fixes.)

**Comment by frankduncan** [comment]
I think that allowing a reporter for the default a la sliders is the best user experience (defaulting that to zero), however that is a higher implementation cost, and so may not be worth the effort.  I'm not wedded to any solution, though.  Whoever works on this eventually can make the decision.

**Comment by DenoBeno** [comment]
Dear Seth & Frank,

Thx for your attention. 

The problem is clear now, proposed solutions sound OK. Wish you luck with fixing it. 

There is nothing I can contribute anymore, so I'll leave you in peace now. :-)

regards
 Denis

**Comment by Idloj** [comment]
Isn't this a duplicate of #386?

**Comment by SethTisue** [comment]
Yes, good point. I will close #386, since this ticket has the better discussion thread.

**Comment by mrerrormessage** [comment]
My preferred fix for this is that when an input box is switched to a color or number input, the input box's default value is set to black / 0. This corresponds to the behavior for string-typed input boxes, which default to an empty string, and square well with the behavior of NetLogo globals, which begin defined as 0. This approach would fix the problem for users without changing the NetLogo concept of what an input box is.

Regarding editing: in the case of colors, a user can only choose other valid colors. In the case of numbers, the validation error "not a valid number" should prevent the user from leaving the input box if their text entry is invalid. Perhaps that error includes two options: "Cancel" (e.g. return to input box) and a "Set to 0" option which automatically sets the value to zero.
